File: Problems/Algorithms/1539. Kth Missing Positive Number/KthMissingNumber.kt
class Solution {
    fun findKthPositive(arr: IntArray, k: Int): Int {
        var left = 0
        var right = arr.size
        
        while (left < right) {
            val mid = left + (right - left) / 2
            
            if (arr[mid] - mid - 1 < k) {
                left = mid + 1
            } else {
                right = mid
            }
        }
        
        return left + k
    }
}

File: Problems/Algorithms/1539. Kth Missing Positive Number/kth_missing_number.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        left, right = -1, len(arr)
        while right - left > 1:
            mid = left + (right - left) // 2
            if arr[mid] - 1 - mid < k:
                left = mid
            else:
                right = mid
        return left + k + 1
